src/Tasks.UnitTests/AssemblyDependency/Node/OutOfProcRarNode_Tests.cs

```diff
@@ -0,0 +1,118 @@
+    /// or RAR logic itself here.
+    /// </summary>
+    public sealed class OutOfProcRarNode_Tests(ITestOutputHelper output)
+    {
+        [Fact]
+        public void RunsOutOfProcIfAllFlagsAreEnabled()
+        {
+            MockEngine engine = new(output)
+            {
+                SetIsOutOfProcRarNodeEnabled = true,
+            };
+            ResolveAssemblyReference rar = new()
+            {
+                AllowOutOfProcNode = true,
+                BuildEngine = engine,
+                Assemblies = [new TaskItem("System")],
+                SearchPaths = [Path.GetDirectoryName(typeof(object).Module.FullyQualifiedName)],
+            };
+
+            using OutOfProcRarNodeEndpoint endpoint = new(
+                    endpointId: 0,
+                    OutOfProcRarNodeEndpoint.CreateConfig(maxNumberOfServerInstances: 1));
+            using CancellationTokenSource cts = new();
+            System.Threading.Tasks.Task runTask = endpoint.RunAsync(cts.Token);
+
+            bool result = rar.Execute();
+
+            // If the out-of-proc path was executed, a client should be registered.
+            using OutOfProcRarClient? rarClient = engine.GetRegisteredTaskObject(OutOfProcRarClient.TaskObjectCacheKey, RegisteredTaskObjectLifetime.Build) as OutOfProcRarClient;
+            Assert.NotNull(rarClient);
+            Assert.True(result);
+            Assert.Equal(0, engine.Warnings);
+            Assert.Equal(0, engine.Errors);
+            _ = Assert.Single(rar.ResolvedFiles);
+
+            rarClient.Dispose();
+            cts.Cancel();
+            runTask.GetAwaiter().GetResult();
+        }
+
+        [Theory]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(false, false)]
+        public void SkipsOutOfProcNodeIfAnyFlagIsDisabled(bool buildEngineFlag, bool taskInputFlag)
+        {
+            MockEngine engine = new(output)
+            {
+                SetIsOutOfProcRarNodeEnabled = buildEngineFlag,
+            };
+            ResolveAssemblyReference rar = new()
+            {
+                AllowOutOfProcNode = taskInputFlag,
+                BuildEngine = engine,
+                Assemblies = [new TaskItem("System")],
+                SearchPaths = [Path.GetDirectoryName(typeof(object).Module.FullyQualifiedName)],
+            };
+
+            bool result = rar.Execute();
+
+            // If the out-of-proc path was skipped, no client should be registered.
+            using OutOfProcRarClient? rarClient = engine.GetRegisteredTaskObject(OutOfProcRarClient.TaskObjectCacheKey, RegisteredTaskObjectLifetime.Build) as OutOfProcRarClient;
+            Assert.Null(rarClient);
+            Assert.True(result);
+            Assert.Equal(0, engine.Warnings);
+            Assert.Equal(0, engine.Errors);
+            _ = Assert.Single(rar.ResolvedFiles);
+        }
+
+        [Fact]
+        public void FallsBackToInProcTaskIfClientFailsToConnect()
```

Comment:
How does this guarantee the client fails to connect?

Reply:
The test sets all the params for RAR to run OOP but never creates the pipe server. The setup is otherwise identical to `RunsOutOfProcIfAllFlagsAreEnabled`, so that's the only way this should fall back.